src/PrettyPrinting.jl

```diff
@@ -1382,150 +1382,245 @@ end
+    If the object is stored in several variable, the first one will be used.
+
+For this to work in doctests, one should call
+`AbstractAlgebra.set_current_module(@__MODULE__)` in the `value` argument of 
+`Documenter.DocMeta.setdocmeta!` and keep the default value of `M = Main` here.
+
+!!! warning
+    This function should not be used directly, but rather through [`AbstractAlgebra.get_name`](@ref).
+"""
+function find_name(obj, M=Main; all::Bool=false)
   # in Documenter, the examples are not run in the REPL.
   # in the REPL: A = ... adds A to the global name space (Main....)
   # in Documenter (doctests) all examples are run in their own module
   # which is stored in CurrentModule, hence we need to search there as well
   #furthermore, they are not exported, hence the "all" option
-  if M === Main && AbstractAlgebra.get_current_module() != Main
-    a = find_name(A, AbstractAlgebra.get_current_module(), all = true)
-    if a !== nothing
+  if M === Main && get_current_module() != Main
+    a = find_name(obj, get_current_module(), all=true)
+    if !isnothing(a)
       return a
     end
   end
-  for a = names(M, all = all)
+  for a = names(M; all=all)
     a === :ans && continue
-    if isdefined(M, a) && getfield(M, a) === A
-        return a
+    if isdefined(M, a) && getfield(M, a) === obj
+      return string(a)
     end
   end
 end
 
-macro show_special(io, O)
-  return :( begin
-    local i = $(esc(io))
-    local o = $(esc(O))
-    s = get_attribute(o, :show)
-    if s !== nothing
-      s(i, o)
-      return
+"""
+   get_name(obj)
+
+Returns the name of the object `obj` if it is set, or `nothing` otherwise.
+This function tries to find a name in the following order:
+1. The name set by [`AbstractAlgebra.set_name!`](@ref).
+2. The name of a variable in global (`Main` module) namespace with value bound to the object `obj` (see [`AbstractAlgebra.find_name`](@ref)).
+3. The name returned by [`AbstractAlgebra.extra_name`](@ref).
+"""
+function get_name(obj)
+  name = nothing
+  if AbstractAlgebra._is_attribute_storing_type(typeof(obj))
+   name = get_attribute(obj, :name)
+  end
+  if isnothing(name)
+    sy = find_name(obj)
+    if !isnothing(sy)
+      name = string(sy)
     end
-  end )
+  end
+  if isnothing(name)
+    name_maybe = extra_name(obj)::Union{String,Nothing}
+    if !isnothing(name_maybe)
+      name = name_maybe
+    end
+  end
+  return name
 end
 
-macro show_special(io, mime, O)
-  return :( begin
-    local i = $(esc(io))
-    local m = $(esc(mime))
-    local o = $(esc(O))
-    s = get_attribute(o, :show)
-    if s !== nothing
-      s(i, m, o)
-      return
+"""
+    @show_name(io, obj)
+
+If compact or supercompact printing is enabled,
+prints the name [`get_name(obj)`](@ref AbstractAlgebra.get_name) of the object `obj` to the `io` stream.
+This macro either prints the name and returns from the current scope, or does nothing.
+
+It is expected to be used in the `show` method of a type shown in the documentation.
```

Comment:
I don't understand the significance "of a type" and "shown in the documentation". Can't we just use the following (if not: what is missing that we would like the user to understand):
```suggestion
It is supposed to be used at the start of `show` methods.
```